app/src/Controller/CoordinatesBattleController.php

```diff
@@ -62,55 +64,71 @@ public function battle()
     public function getNewCoordinate()
     {
         if ($this->request->is('post')) {
-            $like_coordinate_id = filter_input(INPUT_POST, 'liked_coordinate_id', FILTER_SANITIZE_NUMBER_INT);
-            $dislike_coordinate_id = filter_input(INPUT_POST, 'disliked_coordinate_id', FILTER_SANITIZE_NUMBER_INT);
-            if ($like_coordinate_id === NULL || $like_coordinate_id === false
-                || $dislike_coordinate_id === NULL || $dislike_coordinate_id === false
+            $like_coordinate_id = filter_input(
+                INPUT_POST, 'liked_coordinate_id', FILTER_SANITIZE_NUMBER_INT
+            );
+            $dislike_coordinate_id = filter_input(
+                INPUT_POST, 'disliked_coordinate_id', FILTER_SANITIZE_NUMBER_INT
+            );
+            if ($like_coordinate_id === null || $like_coordinate_id === false
+                || $dislike_coordinate_id === null || $dislike_coordinate_id === false
             ) {
                 error_log('Illegal value type');
-                echo '{"hasSucceeded": false}';
+                echo sprintf(
+                    '{"hasSucceeded":false, "errorMessage":"%s"}',
+                    "POSTメソッドで送信された値が不正でした．" .
+                    "ページをリロードするか，ヘッダーの「Unichronicle」からTOPページへ" .
+                    "戻り，Play しなおしてください(これまでのバトル経過は破棄されます)．"
+                );
                 exit;
             }
+            $criteria_json_string = $this->request->data('coordinate_criteria');
 
-            $like_coordinate = $this->Coordinates->get($like_coordinate_id);
-            $like_coordinate->n_like = $like_coordinate->n_like + 1;
-            $this->Coordinates->save($like_coordinate);
-
-            $dislike_coordinate = $this->Coordinates->get($dislike_coordinate_id);
-            $dislike_coordinate->n_unlike = $dislike_coordinate->n_unlike + 1;
-            $this->Coordinates->save($dislike_coordinate);
+            $this->_incrementCoordinatesPoint('n_like', $like_coordinate_id);
+            $this->_incrementCoordinatesPoint('n_unlike', $dislike_coordinate_id);
 
-            $duplicated_flg = false;
             $n_loop = 0;
+            $err_message = sprintf(
+                '{"hasSucceeded":false, "errorMessage":"%s"}',
+                "条件に一致するコーデが存在しません"
+            );
             while (true) {
-                $coordinates = $this->Coordinates->find(
-                    'all',
-                    [
-                        'order' => 'rand()',
-                        'limit' => 1
-                    ]
-                );
+                // 条件にあったコーデ群を取得する
+                // 検索条件毎に一意の文字列(JSON文字列に対してSHA-1を使用する)を生成し，
+                // それをキーとしてキャッシュする
+                $filtered_coordinates = $this->_createQueryFromCriteriaJson(
+                    $criteria_json_string
+                )->cache('filtered_coordinates_' . sha1($criteria_json_string));
```

Comment:
sha1かけてるのは，キーとしてJSONの文字列そのまま後ろにくっつけとくのはかっこわるいかなぁと思ったからなんだけど，必要ないかも

```php
/**
 * Enable result caching for this query.
 *
 * If a query has caching enabled, it will do the following when executed:
 *
 * - Check the cache for $key. If there are results no SQL will be executed.
 *   Instead the cached results will be returned.
 * - When the cached data is stale/missing the result set will be cached as the query
 *   is executed.
 * ...
 */
public function cache($key, $config = 'default')
    {
```
コメント読むとそういうことらしい :smiley: